software/contrib/sequential_switch.py

```diff
@@ -0,0 +1,319 @@
+"""Sequential switch for the EuroPi
+
+    @author Chris Iverach-Brereton <ve4cib@gmail.com>
+    @date   2023-02-13
+"""
+
+from europi import *
+from europi_script import EuroPiScript
+import time
+import random
+
+## Whe in triggered mode we only quantize when we receive an external clock signal
+MODE_SEQUENTIAL=0
+
+## Same as MODE_SEQUENTIAL, but goes backwards
+MODE_REVERSE=1
+
+## Goes back and forth through the outputs
+MODE_PINGPONG=2
+
+## In continuous mode the digital input is ignored and we quantize the input
+#  at the highest rate possible
+MODE_RANDOM=3
+
+## How many milliseconds of idleness do we need before we trigger the screensaver?
+#
+#  =20 minutes
+SCREENSAVER_TIMEOUT_MS = 1000 * 60 * 20
+
+def linear_rescale(x, old_min, old_max, new_min, new_max, clip=True):
+    """Convert a number in one range to another
+
+    @param x        The value to convert
+    @param old_min  The old minimum value for x
+    @param old_max  The old maximum value for x
+    @param new_min  The new minimum value for x
+    @param new_max  The new maximum value for x
+    @param clip     If true, output is always between max and min. Otherwise it's extrapolated
+    @return x, linearly shifted to lie on the new scale
+    """
+    if x < old_min and clip:
+        return new_min
+    elif x > old_max and clip:
+        return new_max
+    else:
+        return (x-old_min) / (old_max-old_min) * (new_max - new_min) + new_min
+
+def knob_rescale(knob, new_min, new_max):
+    """Wrapper for linear_rescale specifically for scaling knob values
+    @param knob  Either europi.k1 or europi.k2; the knob whose value we'll read and rescale
+    @param new_min  The new minimum value for x
+    @param new_max  The new maximum value for x
+    @return The knob's current position, linearly rescaled to lie between new_min and new_max
+    """
+    return linear_rescale(knob.read_position(), 0, 100, new_min, new_max, True)
+    
+
+class ScreensaverScreen:
+    """Blank the screen when idle
+    Eventually it might be neat to have an animation, but that's
+    not necessary for now
+    """
+    
+    def __init__(self, parent):
+        self.parent = parent
+        
+    def on_button1(self):
+        self.parent.active_screen = self.parent.switch_screen
+        self.parent.on_trigger()
+    
+    def draw(self):
+        oled.fill(0)
+        oled.show()
+
+class SwitchScreen:
```

Comment:
[discussion] I really like these screen classes that you use to organize your UI, it ends up being nice and clean. I could see this style getting abstracted a little bit and pulled up into the firmware somewhere.

Reply:
Because the quantizer, sequential switch, and logic programs are all in separate PRs I didn't do much work to abstract out a lot of this. But it's definitely on my list to make some more modular, re-usable UI classes eventually.